### PR TITLE
bpo-47120: define virtual opcodes in opcode.py. Better organization of jump opcodes.

### PR DESCRIPTION
**Pull request by iritkatriel**
https://bugs.python.org/issue47120

### CODE REVIEW

File: Lib/opcode.py [hunk 1]
```diff
@@ -31,28 +31,47 @@
...
 hasfree = []
 hasnargs = [] # unused
+assembler = []
 
 opmap = {}
 opname = ['<%r>' % (op,) for op in range(256)]
+virtual = {}
+
+RELATIVE = 0x1
+ABSOLUTE = 0x2
```
Comment (markshannon) [inline Lib/opcode.py:41]: Aren't we about to remove all absolute jumps?

File: Lib/opcode.py [hunk 2]
```diff
@@ -197,8 +205,32 @@ def jabs_op(name, op, entries=0):
...
+jump_op(ABSOLUTE, 'POP_JUMP_IF_TRUE', 181)
+jump_op(ABSOLUTE, 'POP_JUMP_IF_NOT_NONE', 182)
+jump_op(ABSOLUTE, 'POP_JUMP_IF_NONE', 183)
+
+
+# Virtual Opcodes
+# (Pseudo-instructions used in the compiler,
+# but turned into NOPs by the assembler)
+
+def_op('SETUP_FINALLY', -1)
```
Comment (markshannon) [inline Lib/opcode.py:226]: These are an internal detail of the compiler. They never exist in the final bytecode, so I don't think they should be here.

**Review by markshannon — COMMENTED** [review]
I'm not keen on this change.
It exposes a lot of highly changeable, internal details.

The "virtual" instructions don't exist outside of the compiler, so why expose them?
And why renumber all the jumps?

File: Lib/opcode.py (hunk 2 above)
Comment (iritkatriel) [reply Lib/opcode.py:226]: The point was to autogenerate some of the macros around them.

File: Lib/opcode.py (hunk 1 above)
Comment (iritkatriel) [reply Lib/opcode.py:41]: Yes, and then this flag will not be needed.

**Comment by iritkatriel** [comment]
> And why renumber all the jumps?

We will need to do that if we want to use the evenness of a jump opcode to determine its direction (I'm not entirely sure we want to do that though).

File: Lib/opcode.py (hunk 2 above)
Comment (markshannon) [reply Lib/opcode.py:226]: If the code to generate the macro is large than macros, that suggests to me that it isn't worth it.

Comment (iritkatriel) [reply Lib/opcode.py:226]: For me it's not about how long the code is, but about how many places we need to change when we add/remove/change an opcode (i.e., maintainability).

**Comment by iritkatriel** [comment]
I'll make the jumps change with the current scheme and we can refactor afterwards when the mess is more obvious.

**Comment by markshannon** [comment]
Ultimately I expect that we will want all opcodes to be auto-generated from profiling information, to reduce the effective size of the dispatch table.

**Pull request closed without merging**